### PR TITLE
Theano Support

### PR DESCRIPTION
**Pull request by alecGraves**
I got ```yad2k.py```, ```test_yolo.py```, and the training part of ```retrain_yolo.py``` to work using theano backend.

The main issue I have come across is converting the output features to final boxes without using the tf session. Also, I could not figure out a way to make non max supression elegantly using the keras backend, so I ended up using a [python/numpy implementation](https://github.com/allanzelener/YAD2K/pull/47/files#diff-e3ae7ac93e96dbaeb2761c2f9007fcf3R414) someone else made. After doing that, I just converted most of the final box processing code to numpy.

I am not sure if this is a path you want to take, as using numpy to do the final box processing may not be as fast as using custom layers.

This breaks some stuff, so It is not ready to merge yet. The final box prediction parts of ```retrain_yolo.py``` need to be updated, and I did not touch the ```train_overfit.py``` demo.

Mostly, I just wanted to share my progress and get some feedback. I did this because I needed to run this on ARM, and TensorFlow does not have support for that architecture.

### CODE REVIEW

**Comment by shadySource** [comment]
Addresses #1

**Comment by allanzelener** [comment]
Hi shadySource, thanks for this comprehensive PR that hits a lot of the todos I've put off!

I've been super busy lately (just moved and got a new job) and so it will probably take me another few weeks to get around to reviewing and testing your PR. But giving it a quick look, it looks quite nice and like what I had in mind.

**Comment by alecGraves** [comment]
~~The changes I made seem to have made running the model 3x slower (probably due to converting all of the post-processing to numpy). It could just be how I set up the [final evaluation](https://github.com/shadySource/YAD2K/blob/0793a7665d490ae77ede2775e26ff1394c3d623a/retrain_yolo.py#L311), but converting final processing to a backend/layers is probably a good idea.~~

edit: I was building a new computation graph and evaluating it for every image, which takes 0.2 seconds. To avoid this, I made yolo_head_np(), which is about 200x faster than the silliness I was doing before.

**Comment by alecGraves** [comment]
Still need to fix the paths in retrain_yolo so it does not look for model stuff in cwd. Besides this, it seems pretty ok. 
Edit: Documentation for retraining should be added to the readme, and switching back to hdf5 for training data might be good. 

(Future) Adding support for fit generator with full data augmentation for boxes would be cool. (right after that custom loss function)

File: yad2k/models/keras_yolo.py [hunk 1]
```diff
@@ -51,19 +98,15 @@ def yolo_body(inputs, num_anchors, num_classes):
...
-    x = concatenate([conv21_reshaped, conv20])
+    conv21_reshaped = SpaceToDepth(block_size=2)(conv21)
+    x = Concatenate()([conv21_reshaped, conv20])
     x = DarknetConv2D_BN_Leaky(1024, (3, 3))(x)
     x = DarknetConv2D(num_anchors * (num_classes + 5), (1, 1))(x)
     return Model(inputs, x)
 
 
-def yolo_head(feats, anchors, num_classes):
+def yolo_head(feats, anchors, num_classes): # TODO: Turn this into a layer?
```
Comment (PavlosMelissinos) [inline yad2k/models/keras_yolo.py:109]: You could use this if you want.
https://gist.github.com/PavlosMelissinos/162621051c906ea85b772997d982403a
